src/wfgoto.cpp

```diff
@@ -95,61 +96,55 @@ bool CompareNodes(const PDNODE& a, const PDNODE& b)
 	return ParentOrdering(a, b) < 0;
 }
 
-vector<PDNODE> *FilterBySubtree(vector<PDNODE> * parents, vector<PDNODE> * children)
+vector<PDNODE> FilterBySubtree(vector<PDNODE> const& parents, vector<PDNODE>  const& children)
 {
-	if (parents == NULL || children == NULL)
-		return NULL;
-
-	vector<PDNODE> *results = new vector <PDNODE>();
+	vector<PDNODE> results;
 
 	// for each child, if parent in parents, return
-	for (int i = 0; i < children->size(); i++)
+	std::copy_if(children.begin(),
+				 children.end(),
+				 std::back_inserter(results),
+				 [&parents](PDNODE const& child)
 	{
-		PDNODE child = children->at(i);
 		PDNODE parent = child->pParent;
-
-		if (find(parents->begin(), parents->end(), parent) != parents->end())
-		{
-			results->insert(results->end(), child);
-		}
-	}
+		return (find(parents.begin(), parents.end(), parent) != parents.end());
+	});
 
 	return results;
 }
 
-vector<PDNODE> *TreeIntersection(vector<vector<PDNODE> *> trees)
+vector<PDNODE> TreeIntersection(vector<vector<PDNODE>>& trees)
 {
+	vector<PDNODE> result;
 	int count = trees.size();
 
 	if (count == 0)
-		return NULL;
+		return result;
+
+	// If any tree is empty, return empty
+	if (std::any_of(trees.begin(), trees.end(), [](auto& tree) { return tree.size() == 0; }))

```

Comment:
Same as above concerning but `std::cbegin` and `std::cend` instead since this can be const. Also shouldn't it be `const auto&` and use `tree.empty()`?